include/deal.II/base/mpi.h

```diff
@@ -952,6 +967,28 @@ namespace Utilities
 #  endif
     }
 
+
+#  ifdef DEAL_II_WITH_MPI
+    template <class Iterator, typename Number>
+    std::pair<Number, typename numbers::NumberTraits<Number>::real_type>
+    mean_std(const Iterator begin, const Iterator end, const MPI_Comm &comm)
+    {
+      using Std        = typename numbers::NumberTraits<Number>::real_type;
+      const Number sum = std::accumulate(begin, end, Number(0.));
+
+      const auto size =
+        static_cast<Std>(Utilities::MPI::sum(std::distance(begin, end), comm));
+      Assert(size > 0, ExcDivideByZero());
+      const Number mean   = Utilities::MPI::sum(sum, comm) / size;
+      Std          sq_sum = 0.;
+      std::for_each(begin, end, [&mean, &sq_sum](const Number &v) {
+        sq_sum += numbers::NumberTraits<Number>::abs_square(v - mean);
+      });
+      sq_sum = Utilities::MPI::sum(sq_sum, comm);
```

Comment:
This is somewhat sub-optimal in the sense that it uses three calls to `MPI_Allreduce` rather than one. Well, one summation requires a bit of work because one needs some incremental formula of the standard deviation (there is one in http://dx.doi.org/10.1145/2807591.2807644 section 3.1.2) and a custom MPI allreduce function. Since this function is not expected to be on the critical path, I'd say we keep things as they are, to have the most simple code.

Reply:
> Since this function is not expected to be on the critical path, I'd say we keep things as they are, to have the most simple code.

yes, I also expected this to be used just for post-processing some data, so I did even research how to optimize MPI communication.

Reply:
Yes, apart from the communication issue, there are "online formulas" that only do one pass over all data and that are numerically more stable. It's funny because I just researched what algorithms there are a couple of weeks ago, and here's my summary: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm and https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Online .

I'm not suggesting that you implement these one-pass algorithms, but I think you ought to document the formula you use in the documentation of the function, and possibly leave a note with links as above to possible improvements. I'm saying this because I've learned that with everything statistics, it's so easy to get wrong answers and documenting what you do in a concise way is a good way to help users figure out what they can expect.